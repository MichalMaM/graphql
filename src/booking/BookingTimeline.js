// @flow

import type { DepartureArrival, Leg as LegType } from '../flight/Flight';

export type BookingTimelineEvent =
  | BookedFlightTimelineEvent
  | BookingConfirmedTimelineEvent
  | PaymentConfirmedTimelineEvent
  | DownloadInvoiceTimelineEvent
  | DownloadETicketTimelineEvent
  | DownloadBoardingPassTimelineEvent
  | LeaveForAirportTimelineEvent
  | AirportArrivalTimelineEvent
  | BoardingTimelineEvent
  | DepartureTimelineEvent
  | ArrivalTimelineEvent
  | TransportFromAirportTimelineEvent
  | NoMoreEditsTimelineEvent
  | BagDropTimelineEvent
  | BagPickupTimelineEvent;

export type BookingTimelineData = {|
  events: $ReadOnlyArray<BookingTimelineEvent>,
|};

export type BookedFlightTimelineEvent = {|
  +timestamp: Date,
  +type: 'BookedFlightTimelineEvent',
  +arrival: DepartureArrival,
|};

export type BookingConfirmedTimelineEvent = {|
  +timestamp: Date,
  +type: 'BookingConfirmedTimelineEvent',
|};

export type PaymentConfirmedTimelineEvent = {|
  +timestamp: Date,
  +type: 'PaymentConfirmedTimelineEvent',
|};

export type DownloadInvoiceTimelineEvent = {|
  +timestamp: Date,
  +type: 'DownloadInvoiceTimelineEvent',
  +invoiceUrl: ?string,
  +numberPassengers: number,
  +legs: LegType[],
|};

export type DownloadETicketTimelineEvent = {|
  +timestamp: Date,
  +type: 'DownloadETicketTimelineEvent',
  +ticketUrl: ?string,
|};

export type DownloadBoardingPassTimelineEvent = {|
  +timestamp: Date,
  +type: 'DownloadBoardingPassTimelineEvent',
  +leg: LegType,
|};

export type LeaveForAirportTimelineEvent = {|
  +timestamp: Date,
  +type: 'LeaveForAirportTimelineEvent',
|};

export type AirportArrivalTimelineEvent = {|
  +timestamp: Date,
  +type: 'AirportArrivalTimelineEvent',
  +departure: DepartureArrival,
|};

export type BoardingTimelineEvent = {|
  +timestamp: Date,
  +type: 'BoardingTimelineEvent',
  +gate: ?string,
|};

export type DepartureTimelineEvent = {|
  +timestamp: Date,
  +type: 'DepartureTimelineEvent',
  +arrival: DepartureArrival,
  +duration: ?number,
  +airlineCode: string,
  +flightNumber: number,
|};

export type ArrivalTimelineEvent = {|
  +timestamp: Date,
  +type: 'ArrivalTimelineEvent',
  +arrival: DepartureArrival,
|};

export type TransportFromAirportTimelineEvent = {|
  +timestamp: Date,
  +type: 'TransportFromAirportTimelineEvent',
|};

export type NoMoreEditsTimelineEvent = {|
  +timestamp: Date,
  +type: 'NoMoreEditsTimelineEvent',
|};

export type BagDropTimelineEvent = {|
  +timestamp: Date,
  +type: 'BagDropTimelineEvent',
|};

export type BagPickupTimelineEvent = {|
  +timestamp: Date,
  +type: 'BagPickupTimelineEvent',
|};
