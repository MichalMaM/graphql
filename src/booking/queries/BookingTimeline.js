// @flow

import { GraphQLNonNull, GraphQLID } from 'graphql';
import { fromGlobalId } from 'graphql-relay';

import GraphQLBookingTimeline from '../types/outputs/BookingTimeline';
import type { GraphqlContextType } from '../../common/services/GraphqlContext';
import type { BookingTimelineData } from '../BookingTimeline';

import generateEventsFrom from '../resolvers/BookingTimeline';

export default {
  type: GraphQLBookingTimeline,
  description: 'Get the timeline for your booking.',
  args: {
    id: {
      type: new GraphQLNonNull(GraphQLID),
      description: 'Only Booking id can be used as ID.',
    },
  },
  resolve: async (
    ancestor: mixed,
    { id, brand }: Object,
    { dataLoader }: GraphqlContextType,
  ): Promise<BookingTimelineData> => {
    const { id: originalId, type } = fromGlobalId(id);

    if (
      ![
        'Booking',
        'BookingOneWay',
        'BookingReturn',
        'BookingMulticity',
      ].includes(type)
    ) {
      throw new Error(
        `Booking ID mishmash. You cannot query Booking with ID ` +
          `'${id}' because this ID is not ID of the Booking. ` +
          `Please use opaque ID of the Booking.`,
      );
    }

    const booking = await dataLoader.booking.load({id: originalId, brand});

    const events = generateEventsFrom(booking);

    return { events };
  },
};
