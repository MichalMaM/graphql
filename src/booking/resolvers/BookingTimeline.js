// @flow

import idx from 'idx';
import { DateTime } from 'luxon';

import getFlightDurationInMinutes from '../../flight/resolvers/FlightDuration';

import type {
  BookingTimelineEvent,
  BookedFlightTimelineEvent as BookedFlightType,
  BookingConfirmedTimelineEvent as BookingConfirmedType,
  PaymentConfirmedTimelineEvent as PaymentConfirmedType,
  DownloadInvoiceTimelineEvent as DownloadInvoiceType,
  DownloadETicketTimelineEvent as DownloadETicketType,
  DownloadBoardingPassTimelineEvent as DownloadBoardingPassType,
  LeaveForAirportTimelineEvent as LeaveForAirportType,
  AirportArrivalTimelineEvent as AirportArrivalType,
  BoardingTimelineEvent as BoardingType,
  DepartureTimelineEvent as DepartureType,
  ArrivalTimelineEvent as ArrivalType,
  TransportFromAirportTimelineEvent as TransportFromAirportType,
  NoMoreEditsTimelineEvent as NoMoreEditsType,
  BagDropTimelineEvent as BagDropType,
  BagPickupTimelineEvent as BagPickupType,
} from '../BookingTimeline';
import type { Booking } from '../Booking';
import type { Leg } from '../../flight/Flight';
import type { TripData } from '../types/outputs/Trip';

export default function generateEventsFrom(
  booking: Booking,
): $ReadOnlyArray<BookingTimelineEvent> {
  const numberCheckedBaggage = booking.numberCheckedBaggage || 0;

  const events = [];

  const bookedFlightEvent = generateBookedFlightEvent(booking);
  const bookingConfirmedEvent = generateBookingConfirmedEvent(booking);
  const paymentConfirmedEvent = generatePaymentConfirmedEvent(booking);
  const downloadInvoiceEvent = generateDownloadInvoiceEvent(booking);
  const downloadETicketEvent = generateDownloadETicketEvent(booking);
  const noMoreEditsEvent = generateNoMoreEditsEvent(booking);

  if (bookedFlightEvent) {
    events.push(bookedFlightEvent);
  }
  if (bookingConfirmedEvent) {
    events.push(bookingConfirmedEvent);
  }
  if (paymentConfirmedEvent) {
    events.push(paymentConfirmedEvent);
  }
  if (downloadInvoiceEvent) {
    events.push(downloadInvoiceEvent);
  }
  if (downloadETicketEvent) {
    events.push(downloadETicketEvent);
  }
  if (noMoreEditsEvent) {
    events.push(noMoreEditsEvent);
  }

  let tripEvents = [];
  switch (booking.type) {
    case 'BookingOneWay': {
      const { departure, arrival, legs } = booking;
      tripEvents = generateTripEvents(
        { departure, arrival, legs },
        numberCheckedBaggage,
      );
      break;
    }
    case 'BookingReturn': {
      const { outbound, inbound } = booking;
      tripEvents = [
        ...generateTripEvents(outbound, numberCheckedBaggage),
        ...generateTripEvents(inbound, numberCheckedBaggage),
      ];
      break;
    }
    case 'BookingMulticity': {
      const { trips } = booking;
      trips &&
        trips.forEach(trip =>
          tripEvents.push(...generateTripEvents(trip, numberCheckedBaggage)),
        );
      break;
    }
  }
  events.push(...tripEvents);

  return events;
}

function generateTripEvents(
  trip: ?TripData,
  numberCheckedBaggage: number,
): Array<any> {
  const tripEvents = [];
  if (!trip) {
    return [];
  }
  const leaveForAirportEvent = generateLeaveForAirportEvent(trip);
  if (leaveForAirportEvent) {
    tripEvents.push(leaveForAirportEvent);
  }

  const airportArrivalEvent = generateAirportArrivalEvent(trip);
  if (airportArrivalEvent) {
    tripEvents.push(airportArrivalEvent);
  }

  if (numberCheckedBaggage > 0) {
    const bagDropEvent = generateBagDropEvent(trip);
    if (bagDropEvent) {
      tripEvents.push(bagDropEvent);
    }
  }

  const { legs } = trip;
  if (legs) {
    legs.forEach((leg, index) => {
      if (index !== 0 && numberCheckedBaggage > 0 && leg.recheckRequired) {
        const bagDropEvent = generateBagDropEvent(leg);
        if (bagDropEvent) {
          tripEvents.push(bagDropEvent);
        }
      }

      const boardingEvent = generateBoardingEvent(leg);
      if (boardingEvent) {
        tripEvents.push(boardingEvent);
      }

      const downloadBoardingPassEvent = generateDownloadBoardingPassEvent(leg);
      if (downloadBoardingPassEvent) {
        tripEvents.push(downloadBoardingPassEvent);
      }

      const departureEvent = generateDepartureEvent(leg);
      if (departureEvent) {
        tripEvents.push(departureEvent);
      }

      const arrivalEvent = generateArrivalEvent(leg);
      if (arrivalEvent) {
        tripEvents.push(arrivalEvent);
      }

      if (
        index !== legs.length - 1 &&
        numberCheckedBaggage > 0 &&
        legs[index + 1].recheckRequired
      ) {
        const bagPickupEvent = generateBagPickupEvent(leg);
        if (bagPickupEvent) {
          tripEvents.push(bagPickupEvent);
        }
      }
    });
  }

  if (numberCheckedBaggage > 0) {
    const bagPickupEvent = generateBagPickupEvent(trip);
    if (bagPickupEvent) {
      tripEvents.push(bagPickupEvent);
    }
  }

  const transportFromAirportEvent = generateTransportFromAirportEvent(trip);
  if (transportFromAirportEvent) {
    tripEvents.push(transportFromAirportEvent);
  }

  return tripEvents;
}

export function generateBookedFlightEvent(booking: Booking): BookedFlightType {
  return {
    timestamp: booking.created,
    type: 'BookedFlightTimelineEvent',
    arrival: booking.arrival,
  };
}

export function generateBookingConfirmedEvent(
  booking: Booking,
): ?BookingConfirmedType {
  if (booking.status === 'confirmed') {
    return {
      timestamp: booking.created,
      type: 'BookingConfirmedTimelineEvent',
    };
  }
  return null;
}

export function generatePaymentConfirmedEvent(
  booking: Booking,
): ?PaymentConfirmedType {
  if (booking.status) {
    return {
      timestamp: booking.created,
      type: 'PaymentConfirmedTimelineEvent',
    };
  }
  return null;
}

export function generateDownloadInvoiceEvent(
  booking: Booking,
): ?DownloadInvoiceType {
  const invoiceUrl = idx(booking, _ => _.assets.invoiceUrl) || null;
  const numberPassengers = idx(booking, _ => _.passengers.length) || 0;
  return {
    timestamp: booking.created,
    type: 'DownloadInvoiceTimelineEvent',
    invoiceUrl: invoiceUrl,
    numberPassengers: numberPassengers,
    legs: booking.legs,
  };
}

export function generateDownloadETicketEvent(
  booking: Booking,
): ?DownloadETicketType {
  const ticketUrl = idx(booking.assets, _ => _.ticketUrl) || null;
  return {
    timestamp: booking.created,
    type: 'DownloadETicketTimelineEvent',
    ticketUrl: ticketUrl,
  };
}

export function generateDownloadBoardingPassEvent(
  leg: Leg,
): ?DownloadBoardingPassType {
  const localDepartureTime = idx(leg, _ => _.departure.when.local);
  if (localDepartureTime) {
    return {
      timestamp: localDepartureTime,
      type: 'DownloadBoardingPassTimelineEvent',
      leg,
    };
  }
  return null;
}

export function generateLeaveForAirportEvent(
  booking: Booking | TripData,
): ?LeaveForAirportType {
  const localDepartureTime = idx(booking.departure, _ => _.when.local);
  if (localDepartureTime) {
    const leaveForAiportTime = DateTime.fromJSDate(localDepartureTime, {
      zone: 'UTC',
    })
      .minus({
        hours: 4,
      })
      .toJSDate();
    return {
      timestamp: leaveForAiportTime,
      type: 'LeaveForAirportTimelineEvent',
    };
  }
  return null;
}

export function generateAirportArrivalEvent(
  booking: Booking | TripData,
): ?AirportArrivalType {
  const localDepartureTime = idx(booking.departure, _ => _.when.local);
  if (localDepartureTime) {
    const AiportArrivalTime = DateTime.fromJSDate(localDepartureTime, {
      zone: 'UTC',
    })
      .minus({
        hours: 2,
      })
      .toJSDate();
    return {
      timestamp: AiportArrivalTime,
      type: 'AirportArrivalTimelineEvent',
      departure: booking.departure,
    };
  }
  return null;
}

export function generateBoardingEvent(leg: Leg): ?BoardingType {
  const localDepartureTime = idx(leg, _ => _.departure.when.local);
  if (localDepartureTime) {
    const BoardingTime = DateTime.fromJSDate(localDepartureTime, {
      zone: 'UTC',
    })
      .minus({
        minutes: 30,
      })
      .toJSDate();
    return {
      timestamp: BoardingTime,
      type: 'BoardingTimelineEvent',
      gate: 'gate number', // @TODO Gate Number does not seem available for now...
    };
  }
  return null;
}

export function generateDepartureEvent(leg: Leg): ?DepartureType {
  const departureTime = idx(leg, _ => _.departure.when.local);
  const duration = getFlightDurationInMinutes(leg.departure, leg.arrival);
  const airlineCode = leg.airlineCode;
  const flightNumber = leg.flightNo;
  if (departureTime) {
    return {
      timestamp: departureTime,
      type: 'DepartureTimelineEvent',
      arrival: leg.arrival,
      duration: duration,
      airlineCode: airlineCode,
      flightNumber: flightNumber,
    };
  }
  return null;
}

export function generateArrivalEvent(leg: Leg): ?ArrivalType {
  const arrivalTime = idx(leg, _ => _.arrival.when.local);
  if (arrivalTime) {
    return {
      timestamp: arrivalTime,
      type: 'ArrivalTimelineEvent',
      arrival: leg.arrival,
    };
  }
  return null;
}

export function generateTransportFromAirportEvent(
  booking: Booking | TripData,
): ?TransportFromAirportType {
  const arrivalTime = idx(booking.arrival, _ => _.when.local);
  if (arrivalTime) {
    const transportFromAirportTime = DateTime.fromJSDate(arrivalTime, {
      zone: 'UTC',
    })
      .plus({ minutes: 15 })
      .toJSDate();
    return {
      timestamp: transportFromAirportTime,
      type: 'TransportFromAirportTimelineEvent',
    };
  }
  return null;
}

export function generateNoMoreEditsEvent(booking: Booking): ?NoMoreEditsType {
  const departureTime = idx(booking.departure, _ => _.when.local);
  const allowedToChangeFlights = idx(booking, _ => _.allowedToChangeFlights);
  if (departureTime && allowedToChangeFlights) {
    const noMoreEditsTime = DateTime.fromJSDate(departureTime, {
      zone: 'UTC',
    })
      .minus({ hours: allowedToChangeFlights })
      .toJSDate();
    return {
      timestamp: noMoreEditsTime,
      type: 'NoMoreEditsTimelineEvent',
    };
  }
  return null;
}

export function generateBagDropEvent(legOrTrip: Leg | TripData): ?BagDropType {
  const departureTime = idx(legOrTrip, _ => _.departure.when.local);
  if (departureTime) {
    const bagDropTime = DateTime.fromJSDate(departureTime, {
      zone: 'UTC',
    })
      .minus({ minutes: 35 })
      .toJSDate();
    return {
      timestamp: bagDropTime,
      type: 'BagDropTimelineEvent',
    };
  }
  return null;
}

export function generateBagPickupEvent(
  legOrTrip: Leg | TripData,
): ?BagPickupType {
  const arrivalTime = idx(legOrTrip, _ => _.arrival.when.local);
  if (arrivalTime) {
    const bagPickupTime = DateTime.fromJSDate(arrivalTime, {
      zone: 'UTC',
    })
      .plus({ minutes: 10 })
      .toJSDate();
    return {
      timestamp: bagPickupTime,
      type: 'BagPickupTimelineEvent',
    };
  }
  return null;
}
