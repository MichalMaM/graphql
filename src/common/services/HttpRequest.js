// @flow

import { URL } from 'url';
import { fetchJson } from './JsonFetcher';

export async function get(
  absoluteApiUrl: string,
  token: ?string,
  requestHeaders?: Object = {},
  brand: ?string,
): Promise<Object> {
  const urlObject = new URL(absoluteApiUrl);

  if (token !== null && token !== undefined) {
    urlObject.searchParams.append('token', token);
  }
  if (brand !== null && brand !== undefined) {
    urlObject.searchParams.append('brand', brand);
  }
  const headers = {
    'X-Client': 'graphql',
    'Content-Type': 'application/json',
    ...requestHeaders,
  };

  return fetchJson(urlObject.toString(), 'GET', { headers });
}

export async function post(
  absoluteApiUrl: string,
  payload: Object,
  requestHeaders?: Object = {},
): Promise<Object> {
  const body = JSON.stringify(payload);
  const headers = {
    'Content-Type': 'application/json',
    'X-Client': 'graphql',
    ...requestHeaders,
  };

  return fetchJson(absoluteApiUrl, 'POST', { body, headers });
}
