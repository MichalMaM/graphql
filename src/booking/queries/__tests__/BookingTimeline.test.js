// @flow

import { graphql, RestApiMock } from '../../../common/services/TestingTools';
import config from '../../../../config/application';
import BookingTimeline from '../BookingTimeline';
import AllBookingsDataset from '../../datasets/AllBookings.json';
import Booking2707251 from '../../datasets/booking-2707251.json';
import Booking7474870 from '../../datasets/booking-7474870.json';
import AirlinesDataset from '../../../flight/datasets/airlines.json';

beforeEach(() => {
  RestApiMock.onGet(config.restApiEndpoint.allBookings).replyWithData(
    AllBookingsDataset,
  );
  RestApiMock.onGet(config.restApiEndpoint.airlines).replyWithData(
    AirlinesDataset,
  );
  RestApiMock.onGet(
    'https://booking-api.skypicker.com/api/v0.1/users/self/bookings/2707251?simple_token=b206db64-718f-4608-babb-0b8abe6e1b9d',
  ).replyWithData(Booking2707251);
  RestApiMock.onGet(
    'https://booking-api.skypicker.com/api/v0.1/users/self/bookings/7474870?simple_token=44c137b3-dca2-48ca-b091-aabfa4b53488',
  ).replyWithData(Booking7474870);

  ['CDG', 'LGW', 'VIE', 'HKG', 'PEK'].forEach(iata => {
    RestApiMock.onGet(
      config.restApiEndpoint.allLocations({
        type: 'id',
        id: iata,
        locale: 'en-US',
      }),
    ).replyWithData({
      locations: [
        {
          id: 'MOCKED',
          city: {
            name: 'Mocked City Name',
          },
        },
      ],
    });
  });
});

describe('single booking timeline query', () => {
  it('should be of BookingTimeline type', () => {
    expect(BookingTimeline.type.toString()).toBe('BookingTimeline');
  });

  it('should not accept database id', async () => {
    const databaseIdQuery = `{
      bookingTimeline(id: 2707251) {
        __typename
      }
    }`;
    expect(await graphql(databaseIdQuery)).toMatchSnapshot();
  });

  it('should work with opaque Booking id', async () => {
    const query = `{
      bookingTimeline(id: "Qm9va2luZzoyNzA3MjUx") {
        events {
          __typename
            timestamp
          ... on DownloadInvoiceTimelineEvent{
            invoiceUrl
            numberPassengers
            legs {
              departure {
                airport {
                  city {
                    name
                  }
                }
              }
              arrival{
                airport {
                  city {
                    name
                  }
                }
              }
            }
          }
          ... on  DownloadETicketTimelineEvent {
            ticketUrl
          }
          ... on  AirportArrivalTimelineEvent {
            location {
              airport {
                city {
                  locationId
                  name
                  slug
                }
              }
            }
          }
          ... on BoardingTimelineEvent {
            gate
          }
          ... on  DepartureTimelineEvent {
            location {
              airport {
                city {
                  name
                }
              }
            }
            duration
            airline {
              code
              name
            }
            flightNumber
          }
          ... on  ArrivalTimelineEvent {
            location {
              airport {
                city {
                  locationId
                  name
                  slug
                }
              }
            }
          }
        }
      }
    }`;
    expect(await graphql(query)).toMatchSnapshot();
  });
  it('should work with another opaque Booking id (has bags)', async () => {
    const query = `{
      bookingTimeline(id: "Qm9va2luZzo3NDc0ODcw4o+OIAo=") {
        events {
          __typename
            timestamp
          ... on DownloadInvoiceTimelineEvent{
            invoiceUrl
            numberPassengers
            legs {
              departure {
                airport {
                  city {
                    name
                  }
                }
              }
              arrival{
                airport {
                  city {
                    name
                  }
                }
              }
            }
          }
          ... on  DownloadETicketTimelineEvent {
            ticketUrl
          }
          ... on  AirportArrivalTimelineEvent {
            location {
              airport {
                city {
                  locationId
                  name
                  slug
                }
              }
            }
          }
          ... on BoardingTimelineEvent {
            gate
          }
          ... on  DepartureTimelineEvent {
            location {
              airport {
                city {
                  name
                }
              }
            }
            duration
            airline {
              code
              name
            }
            flightNumber
          }
          ... on  ArrivalTimelineEvent {
            location {
              airport {
                city {
                  locationId
                  name
                  slug
                }
              }
            }
          }
        }
      }
    }`;
    expect(await graphql(query)).toMatchSnapshot();
  });
});
