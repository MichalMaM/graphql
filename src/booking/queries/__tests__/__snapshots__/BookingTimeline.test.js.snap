// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`single booking timeline query should not accept database id 1`] = `
Object {
  "data": Object {
    "bookingTimeline": null,
  },
  "errors": Array [
    [GraphQLError: Booking ID mishmash. You cannot query Booking with ID '2707251' because this ID is not ID of the Booking. Please use opaque ID of the Booking.],
  ],
}
`;

exports[`single booking timeline query should work with another opaque Booking id (has bags) 1`] = `
Object {
  "data": Object {
    "bookingTimeline": Object {
      "events": Array [
        Object {
          "__typename": "BookedFlightTimelineEvent",
          "timestamp": "2018-07-02T15:04:06.000Z",
        },
        Object {
          "__typename": "BookingConfirmedTimelineEvent",
          "timestamp": "2018-07-02T15:04:06.000Z",
        },
        Object {
          "__typename": "PaymentConfirmedTimelineEvent",
          "timestamp": "2018-07-02T15:04:06.000Z",
        },
        Object {
          "__typename": "DownloadInvoiceTimelineEvent",
          "invoiceUrl": "https://booking.skypicker.com/api/v0.1/get_file/7474870?file_id=603375001&simple_token=44c137b3-dca2-48ca-b091-aabfa4b53488",
          "legs": Array [
            Object {
              "arrival": Object {
                "airport": Object {
                  "city": Object {
                    "name": "Mocked City Name",
                  },
                },
              },
              "departure": Object {
                "airport": Object {
                  "city": Object {
                    "name": "Mocked City Name",
                  },
                },
              },
            },
            Object {
              "arrival": Object {
                "airport": Object {
                  "city": Object {
                    "name": "Mocked City Name",
                  },
                },
              },
              "departure": Object {
                "airport": Object {
                  "city": Object {
                    "name": "Mocked City Name",
                  },
                },
              },
            },
            Object {
              "arrival": Object {
                "airport": Object {
                  "city": Object {
                    "name": "Mocked City Name",
                  },
                },
              },
              "departure": Object {
                "airport": Object {
                  "city": Object {
                    "name": "Mocked City Name",
                  },
                },
              },
            },
            Object {
              "arrival": Object {
                "airport": Object {
                  "city": Object {
                    "name": "Mocked City Name",
                  },
                },
              },
              "departure": Object {
                "airport": Object {
                  "city": Object {
                    "name": "Mocked City Name",
                  },
                },
              },
            },
          ],
          "numberPassengers": 1,
          "timestamp": "2018-07-02T15:04:06.000Z",
        },
        Object {
          "__typename": "DownloadETicketTimelineEvent",
          "ticketUrl": "https://mailing-files-dropzone.s3.eu-west-1.amazonaws.com/7474/1530544118_E-ticket_passenger_e8c265f9a621731d2dc62357d9f55307.pdf?v=1530544121",
          "timestamp": "2018-07-02T15:04:06.000Z",
        },
        Object {
          "__typename": "NoMoreEditsTimelineEvent",
          "timestamp": "2018-08-28T13:30:00.000Z",
        },
        Object {
          "__typename": "LeaveForAirportTimelineEvent",
          "timestamp": "2018-08-30T09:30:00.000Z",
        },
        Object {
          "__typename": "AirportArrivalTimelineEvent",
          "location": Object {
            "airport": Object {
              "city": Object {
                "locationId": null,
                "name": "Mocked City Name",
                "slug": null,
              },
            },
          },
          "timestamp": "2018-08-30T11:30:00.000Z",
        },
        Object {
          "__typename": "BagDropTimelineEvent",
          "timestamp": "2018-08-30T12:55:00.000Z",
        },
        Object {
          "__typename": "BoardingTimelineEvent",
          "gate": "gate number",
          "timestamp": "2018-08-30T13:00:00.000Z",
        },
        Object {
          "__typename": "DownloadBoardingPassTimelineEvent",
          "timestamp": "2018-08-30T13:30:00.000Z",
        },
        Object {
          "__typename": "DepartureTimelineEvent",
          "airline": Object {
            "code": "CA",
            "name": "Air China",
          },
          "duration": 560,
          "flightNumber": 842,
          "location": Object {
            "airport": Object {
              "city": Object {
                "name": "Mocked City Name",
              },
            },
          },
          "timestamp": "2018-08-30T13:30:00.000Z",
        },
        Object {
          "__typename": "ArrivalTimelineEvent",
          "location": Object {
            "airport": Object {
              "city": Object {
                "locationId": null,
                "name": "Mocked City Name",
                "slug": null,
              },
            },
          },
          "timestamp": "2018-08-31T04:50:00.000Z",
        },
        Object {
          "__typename": "BagPickupTimelineEvent",
          "timestamp": "2018-08-31T05:00:00.000Z",
        },
        Object {
          "__typename": "BagDropTimelineEvent",
          "timestamp": "2018-08-31T06:50:00.000Z",
        },
        Object {
          "__typename": "BoardingTimelineEvent",
          "gate": "gate number",
          "timestamp": "2018-08-31T06:55:00.000Z",
        },
        Object {
          "__typename": "DownloadBoardingPassTimelineEvent",
          "timestamp": "2018-08-31T07:25:00.000Z",
        },
        Object {
          "__typename": "DepartureTimelineEvent",
          "airline": Object {
            "code": "CA",
            "name": "Air China",
          },
          "duration": 220,
          "flightNumber": 6533,
          "location": Object {
            "airport": Object {
              "city": Object {
                "name": "Mocked City Name",
              },
            },
          },
          "timestamp": "2018-08-31T07:25:00.000Z",
        },
        Object {
          "__typename": "ArrivalTimelineEvent",
          "location": Object {
            "airport": Object {
              "city": Object {
                "locationId": null,
                "name": "Mocked City Name",
                "slug": null,
              },
            },
          },
          "timestamp": "2018-08-31T11:05:00.000Z",
        },
        Object {
          "__typename": "BagPickupTimelineEvent",
          "timestamp": "2018-08-31T11:15:00.000Z",
        },
        Object {
          "__typename": "TransportFromAirportTimelineEvent",
          "timestamp": "2018-08-31T11:20:00.000Z",
        },
        Object {
          "__typename": "LeaveForAirportTimelineEvent",
          "timestamp": "2018-09-14T15:00:00.000Z",
        },
        Object {
          "__typename": "AirportArrivalTimelineEvent",
          "location": Object {
            "airport": Object {
              "city": Object {
                "locationId": null,
                "name": "Mocked City Name",
                "slug": null,
              },
            },
          },
          "timestamp": "2018-09-14T17:00:00.000Z",
        },
        Object {
          "__typename": "BagDropTimelineEvent",
          "timestamp": "2018-09-14T18:25:00.000Z",
        },
        Object {
          "__typename": "BoardingTimelineEvent",
          "gate": "gate number",
          "timestamp": "2018-09-14T18:30:00.000Z",
        },
        Object {
          "__typename": "DownloadBoardingPassTimelineEvent",
          "timestamp": "2018-09-14T19:00:00.000Z",
        },
        Object {
          "__typename": "DepartureTimelineEvent",
          "airline": Object {
            "code": "CA",
            "name": "Air China",
          },
          "duration": 210,
          "flightNumber": 6516,
          "location": Object {
            "airport": Object {
              "city": Object {
                "name": "Mocked City Name",
              },
            },
          },
          "timestamp": "2018-09-14T19:00:00.000Z",
        },
        Object {
          "__typename": "ArrivalTimelineEvent",
          "location": Object {
            "airport": Object {
              "city": Object {
                "locationId": null,
                "name": "Mocked City Name",
                "slug": null,
              },
            },
          },
          "timestamp": "2018-09-14T22:30:00.000Z",
        },
        Object {
          "__typename": "BoardingTimelineEvent",
          "gate": "gate number",
          "timestamp": "2018-09-15T02:20:00.000Z",
        },
        Object {
          "__typename": "DownloadBoardingPassTimelineEvent",
          "timestamp": "2018-09-15T02:50:00.000Z",
        },
        Object {
          "__typename": "DepartureTimelineEvent",
          "airline": Object {
            "code": "CA",
            "name": "Air China",
          },
          "duration": 600,
          "flightNumber": 841,
          "location": Object {
            "airport": Object {
              "city": Object {
                "name": "Mocked City Name",
              },
            },
          },
          "timestamp": "2018-09-15T02:50:00.000Z",
        },
        Object {
          "__typename": "ArrivalTimelineEvent",
          "location": Object {
            "airport": Object {
              "city": Object {
                "locationId": null,
                "name": "Mocked City Name",
                "slug": null,
              },
            },
          },
          "timestamp": "2018-09-15T06:50:00.000Z",
        },
        Object {
          "__typename": "BagPickupTimelineEvent",
          "timestamp": "2018-09-15T07:00:00.000Z",
        },
        Object {
          "__typename": "TransportFromAirportTimelineEvent",
          "timestamp": "2018-09-15T07:05:00.000Z",
        },
      ],
    },
  },
}
`;

exports[`single booking timeline query should work with opaque Booking id 1`] = `
Object {
  "data": Object {
    "bookingTimeline": Object {
      "events": Array [
        Object {
          "__typename": "BookedFlightTimelineEvent",
          "timestamp": "2017-03-30T07:28:55.000Z",
        },
        Object {
          "__typename": "BookingConfirmedTimelineEvent",
          "timestamp": "2017-03-30T07:28:55.000Z",
        },
        Object {
          "__typename": "PaymentConfirmedTimelineEvent",
          "timestamp": "2017-03-30T07:28:55.000Z",
        },
        Object {
          "__typename": "DownloadInvoiceTimelineEvent",
          "invoiceUrl": null,
          "legs": Array [
            Object {
              "arrival": Object {
                "airport": Object {
                  "city": Object {
                    "name": "Mocked City Name",
                  },
                },
              },
              "departure": Object {
                "airport": Object {
                  "city": Object {
                    "name": "Mocked City Name",
                  },
                },
              },
            },
          ],
          "numberPassengers": 2,
          "timestamp": "2017-03-30T07:28:55.000Z",
        },
        Object {
          "__typename": "DownloadETicketTimelineEvent",
          "ticketUrl": "https://skypicker-mailing.s3.amazonaws.com/2707/1490859068_E-ticket_passenger_and_1_more_e826fc3275a65a5db8808279d8fc7f8f.pdf?v=1490859069",
          "timestamp": "2017-03-30T07:28:55.000Z",
        },
        Object {
          "__typename": "NoMoreEditsTimelineEvent",
          "timestamp": "2017-09-07T21:10:00.000Z",
        },
        Object {
          "__typename": "LeaveForAirportTimelineEvent",
          "timestamp": "2017-09-09T17:10:00.000Z",
        },
        Object {
          "__typename": "AirportArrivalTimelineEvent",
          "location": Object {
            "airport": Object {
              "city": Object {
                "locationId": null,
                "name": "Mocked City Name",
                "slug": null,
              },
            },
          },
          "timestamp": "2017-09-09T19:10:00.000Z",
        },
        Object {
          "__typename": "BoardingTimelineEvent",
          "gate": "gate number",
          "timestamp": "2017-09-09T20:40:00.000Z",
        },
        Object {
          "__typename": "DownloadBoardingPassTimelineEvent",
          "timestamp": "2017-09-09T21:10:00.000Z",
        },
        Object {
          "__typename": "DepartureTimelineEvent",
          "airline": Object {
            "code": "VY",
            "name": "Vueling",
          },
          "duration": 80,
          "flightNumber": 8773,
          "location": Object {
            "airport": Object {
              "city": Object {
                "name": "Mocked City Name",
              },
            },
          },
          "timestamp": "2017-09-09T21:10:00.000Z",
        },
        Object {
          "__typename": "ArrivalTimelineEvent",
          "location": Object {
            "airport": Object {
              "city": Object {
                "locationId": null,
                "name": "Mocked City Name",
                "slug": null,
              },
            },
          },
          "timestamp": "2017-09-09T23:30:00.000Z",
        },
        Object {
          "__typename": "TransportFromAirportTimelineEvent",
          "timestamp": "2017-09-09T23:45:00.000Z",
        },
      ],
    },
  },
}
`;
